### ISSUE TITLE
Cannot actually use the unblock token

### ISSUE THREAD

**Issue by jrgm** [mozilla/fxa-content-server#4324]
STR
1. Create and verify a new account at https://latest.dev.lcip.org/signup?service=sync
2. In a new browser profile, use the email from (1) at https://latest.dev.lcip.org/signin?service=sync, but enter the wrong password. 
3. Repeat #2
4. Repeat #2
5. Repeat #2. This time the UI gives the message about an auth token and and email is sent.
6. Fetch the code from the 'Firefox Account authorization code' email, and enter the code in the UI and submit.

Expected: some benefit from the code.

Actual: UI submits the code and the current wrong password from #5 and shows a failure. Attempting to re-submit with the correct password says 'You've tried too many times. Try again in 15 minutes'.

**Comment by rfk** [comment]
Confirming I can reproduce this.

In the devtools console, I see that the call to /send_unblock_code is being rate-limited along with the call to /login:

![screen shot 2016-10-25 at 14 58 13](https://cloud.githubusercontent.com/assets/34695/19672570/8036a102-9ac3-11e6-9f99-ad1ab088f089.png)

Which IIRC, is rate-limited as an "email sending action".  And indeed, other email-sending actions like password-reset are also rate-limited once I get into this state.

**Comment by rfk** [comment]
Ugh, so I think this is a bug in the way we're counting login failures per IP address.  The code here:

  https://github.com/mozilla/fxa-customs-server/blob/master/lib/ip_record.js#L48

Is *supposed* to "take the highest-weighted bad-login event for each email address", but it will actually sum up multiple bad login events from a single IP address.  So what's happening here appears to be:

* At step number (5), you trip the per-email/ip-pair login rate-limiting.  This blocks any further login attempts, but still lets you send email, so you can get the authorization code.
* At step number (6), entering the authorization code causes the bad login attempt to be tried again.  This trips the buggy per-ip-address rate-limiting, which prevents you from sending emails, so it can't send you any more authorization codes.

**Comment by jrgm** [comment]
Aside from the customs-server change, I'm finding the UX confusing: I get a token, input it and submit... I'd expect to be put back in a "positive" next state. Instead, the UI resubmits my incorrect password and then complains to me that "Incorrect password. To double-check what you've entered, press and hold down the Show button.".  I am left wondering what the whole purpose of that token was.

**Comment by shane-tomlinson** [comment]
> I'd expect to be put back in a "positive" next state. Instead, the UI resubmits my incorrect password and then complains to me that "Incorrect password. 

Signin blocking is meant to prevent attackers from learning they have guessed the user's password. To do so, a check with customs is made before the passwords are compared. If customs returns an "ixnay, you are blocked", we send an email with a code in it and tell the user to enter the code. Checking the password before checking with customs would allow attackers to learn when they have guessed the password.

When the user enters the code, we send the code _along with the username and password_ to /account/login to attempt to sign the user in w/ no additional clicks. This behavior was a request by @ryanfeeley and @davismtl to minimize clicks for the happy path, and it's a nice flow.

I agree the flow is confusing _if the wrong password is typed_. The user is sent back to enter their password again, and then they must enter a new unblock code. Sending a new unblock code is crappy. Maybe to smooth out the flow, an `unblock code` is good for X number of login attempts, where X is some small number. It'd require some additional work on the db, but would definitely help this flow.

cc @seanmonstar.

**Comment by seanmonstar** [comment]
> I'd expect to be put back in a "positive" next state. Instead, the UI resubmits my incorrect password and then complains to me that "Incorrect password. 

Shane did a great job explaining the why. I'd like to just make sure its super clear the what.

1. You are blocked or rate-limited.
2. You enter an email, and the wrong password, and submit.
3. Since you are rate limited, we do not look up the password and compare, but instead offer you an unblock code.
4. Once you've found the unblock code and entered it, we have 2 ways we can go:

----

- Submit the email and password you just submitted, but with the unblock code. Remember, we never checked it before, so we have no idea if it's good or not.
- Ask you to type in your email and password a second time, even though we have never checked it the first time. You'll likely type the exact same thing.

----

That the unblock code is 1-use only pass could be changed. The reasons for its design were: It was far less complicated internally, which meant a) we could build it faster, and b) less complexity means less places to be wrong, which felt important considering this is security related.

Maybe it makes sense to only "consume" the unblock code once it has been properly used. That would leave it valid for your several password attempts (using an unblock code is also tracked separately in customs-server to rate-limit), or till it expires, which is currently configured for 1 hour.

The downside is this means that code is available for an attack for up to an hour, if for some contrived reason they got access to the email account only once. Maybe other downsides, but it does look minimal.

**Comment by davismtl** [comment]
Could the unblock code just reset the password attempts? If we give a user 3 attempts, they fail times, we block them and then use the code... could we not just reset their number of attempts to 3? (independently of the time)

**Comment by seanmonstar** [comment]
There's several reasons a user could be blocked, not just bad password attempts. Status checks are rate-limited, and being on an IP blocklist means blocked. Additionally, attempts are record for an email address, so a user can also be blocked because that specific email has had too many attempts recently, even if those attempts are across the planet (or their neighbor). We would want to allow the user with the unblock code to try again, but not open it up to anyone else that was trying.

**Comment by rfk** [comment]
>  I am left wondering what the whole purpose of that token was.

To @jrgm's point, while we are able to explain what the purpose of the token was, it may not be obvious to the user that this is what happened.  As a user, I successfully did a thing (entered a code from my email) bu then I got an error message.  Did I do it wrong?

I'm going to bikeshed a little below, but my suggestion is that we ship it as-is and gather metrics on how users are hitting it in practice.  If people are mostly entering the right password when unblocking themselves, maybe we don't need to invest in this.  If lots of users are failing out even after unblocking, maybe we invest in this more.

So, a hypothetical experience that might make it more obvious what went wrong:

1. I submit my login credentials, and I get an "authorization code required" form
2. I submit the code from my email, and I see a screen that says "authorization code accepted, logging you in..." with a little spinner
3. I'm bounced back to the login screen with an error message about my password

As an engineer, I know that (2) is completely useless and in fact we'd have to fake it out - we can't know if the code was accepted until after we've tried to log you in.  But as a user, it might re-assure me that I did the code-submitting step right, and that the error is due to something else.

**Comment by vladikoff** [comment]
Assigned to @ryanfeeley

**Comment by vladikoff** [comment]
@ryanfeeley Do we need changes to UX of the unblock flow or is it fine as it is right now?

**Comment by vladikoff** [comment]
Let's reopen or file a new issue if we want to make changes..
